### PR TITLE
More logs

### PR DESCRIPTION
**Pull request by arnaudgiuliani**
Open 0.11.3 fixes + 
Add debug logging + debug logs in critical places

### CODE REVIEW

File: uniflow-android-test/src/main/java/io/uniflow/android/test/SimpleTestObservers.kt [hunk 1]
```diff
@@ -36,7 +35,14 @@ class TestViewObserver {
 
     @Deprecated("better use verifySequence")
     fun assertReceived(vararg any: UIData) = verifySequence(*any)
-    fun verifySequence(vararg any: UIData) = assert(this.values == any.toList()) { "Wrong values\nshould have [${any.toList()}]\nbut was [${values}]" }
+    fun verifySequence(vararg testingData: UIData) {
+        val testingValues = testingData.toList()
+        values.forEachIndexed { index, uiData ->
+            assert(uiData == testingValues[index]) {
+                "Wrong values at [$index] - Type: ${uiData::class.simpleName}\nShould have [${testingValues[index]}]\n  but was [${uiData}]"
+            }
```
Comment (erikhuizinga) [inline uniflow-android-test/src/main/java/io/uniflow/android/test/SimpleTestObservers.kt:43]: Prefer `assertEquals(expected, actual, message)` from package `kotlin.test`.

File: uniflow-android-test/src/main/java/io/uniflow/android/test/SimpleTestObservers.kt [hunk 2]
```diff
@@ -36,7 +35,14 @@ class TestViewObserver {
 
     @Deprecated("better use verifySequence")
     fun assertReceived(vararg any: UIData) = verifySequence(*any)
-    fun verifySequence(vararg any: UIData) = assert(this.values == any.toList()) { "Wrong values\nshould have [${any.toList()}]\nbut was [${values}]" }
+    fun verifySequence(vararg testingData: UIData) {
+        val testingValues = testingData.toList()
+        values.forEachIndexed { index, uiData ->
+            assert(uiData == testingValues[index]) {
+                "Wrong values at [$index] - Type: ${uiData::class.simpleName}\nShould have [${testingValues[index]}]\n  but was [${uiData}]"
```
Comment (erikhuizinga) [inline uniflow-android-test/src/main/java/io/uniflow/android/test/SimpleTestObservers.kt:42]:
```suggestion
                "Wrong value at [$index] - Type: ${uiData::class.simpleName}\nShould have [${testingValues[index]}]\nbut was [$uiData]"
```

File: uniflow-androidx-test/src/main/java/io/uniflow/android/test/SimpleTestObservers.kt [hunk 3]
```diff
@@ -34,7 +33,14 @@ class TestViewObserver {
 
     @Deprecated("better use verifySequence")
     fun assertReceived(vararg any: UIData) = verifySequence(*any)
-    fun verifySequence(vararg any: UIData) = assert(this.values == any.toList()) { "Wrong values\nshould have [${any.toList()}]\nbut was [${values}]" }
+    fun verifySequence(vararg testingData: UIData){
+        val testingValues = testingData.toList()
+        values.forEachIndexed { index, uiData ->
+            assert(uiData == testingValues[index]) {
+                "Wrong values at [$index] - Type: ${uiData::class.simpleName}\nShould have [${testingValues[index]}]\n  but was [${uiData}]"
+            }
```
Comment (erikhuizinga) [inline uniflow-androidx-test/src/main/java/io/uniflow/android/test/SimpleTestObservers.kt:41]: Same comments apply as for uniflow-android-test/src/main/java/io/uniflow/android/test/SimpleTestObservers.kt.

File: uniflow-android/src/main/java/io/uniflow/android/flow/DataFlowObserver.kt [hunk 4]
```diff
@@ -27,9 +29,41 @@ import io.uniflow.core.flow.data.UIState
...
+        event?.let {
+            event.peek().let {
+                UniFlowLogger.debug("onPeekEvents - $this -> $event")
+                handleEvents(it)
+            }
+        }
+    })
+}
+
+fun LifecycleOwner.onTakeEvents(vm: AndroidDataFlow, handleEvents: (UIEvent) -> Unit) {
```
Comment (erikhuizinga) [inline uniflow-android/src/main/java/io/uniflow/android/flow/DataFlowObserver.kt:60]: This function and `onPeekEvents` are new and public, so their behaviour should be documented and they should be added to the changelog.

File: uniflow-core/src/main/kotlin/io/uniflow/core/flow/UIDataStore.kt [hunk 5]
```diff
@@ -11,14 +11,15 @@ class UIDataStore(private val publisher: UIDataPublisher, defaultState: UIState)
...
+        UniFlowLogger.debug("UIDataStore - push: $dataUpdate")
         when (dataUpdate.data) {
             is UIState -> {
                 currentState = dataUpdate.data
                 UniFlowLogger.logState(dataUpdate.data)
                 when (dataUpdate.type) {
                     UIDataUpdateType.PUBLISH -> publisher.publishState(dataUpdate.data)
                     else -> {
-                        UniFlowLogger.log("state notification - ${dataUpdate.data} is not published")
+                        UniFlowLogger.log("[State] internal update '${dataUpdate.data}'")
```
Comment (erikhuizinga) [inline uniflow-core/src/main/kotlin/io/uniflow/core/flow/UIDataStore.kt:22]: Should you use `UniFlowLogger.debug` instead of a regular log?

**Review by erikhuizinga — CHANGES_REQUESTED** [review]
Nice work, some minor comments.

File: uniflow-android-test/src/main/java/io/uniflow/android/test/SimpleTestObservers.kt (hunk 1 above)
Comment (arnaudgiuliani) [reply uniflow-android-test/src/main/java/io/uniflow/android/test/SimpleTestObservers.kt:43]: yep, assert is also from Kotlin no? 🤔

File: uniflow-android/src/main/java/io/uniflow/android/flow/DataFlowObserver.kt (hunk 4 above)
Comment (arnaudgiuliani) [reply uniflow-android/src/main/java/io/uniflow/android/flow/DataFlowObserver.kt:60]: yes, you're right 👍

File: uniflow-core/src/main/kotlin/io/uniflow/core/flow/UIDataStore.kt (hunk 5 above)
Comment (arnaudgiuliani) [reply uniflow-core/src/main/kotlin/io/uniflow/core/flow/UIDataStore.kt:22]: yep, still thinking about that 🤔  I believe it's useful to know the notify thing ... but has to be clear that also that we don't push the state, just the event. But logs are explicit enough

File: uniflow-android-test/src/main/java/io/uniflow/android/test/SimpleTestObservers.kt (hunk 1 above)
Comment (erikhuizinga) [reply uniflow-android-test/src/main/java/io/uniflow/android/test/SimpleTestObservers.kt]: `assert` is a JVM thing. And you have to enable assertions for them to work. By default, assertions are disabled!

- https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/assert.html

Comment (arnaudgiuliani) [reply uniflow-android-test/src/main/java/io/uniflow/android/test/SimpleTestObservers.kt]: 👍

File: gradle/versions.gradle [hunk 6]
```diff
@@ -1,10 +1,10 @@
 ext {
     // Koin
-    uniflow_version = '0.11.2'
+    uniflow_version = '0.11.3'
 
     // Kotlin
-    kotlin_version = '1.3.72'
-    coroutines_version = "1.3.5"
+    kotlin_version = '1.4.10'
+    coroutines_version = "1.3.9"
```
Comment (erikhuizinga) [inline gradle/versions.gradle:7]: Both these version numbers are used for some dependencies that are now `api` (transitive) deps in various modules. This change might affect clients of UniFlow. Therefore, these changes should be documented, at least.

Also, if this project uses [semver](https://semver.org), then maybe the next UniFlow version should be 0.12.0.

File: uniflow-test/build.gradle [hunk 7]
```diff
@@ -5,9 +5,10 @@ description = 'UniFlow a Kotlin unidrectional data flow framework'
 
 dependencies {
     api project(":uniflow-core")
-    implementation "org.jetbrains.kotlinx:kotlinx-coroutines-test:$coroutines_version"
-    implementation "junit:junit:$junit_version"
-    implementation "io.mockk:mockk:$mockk_version"
+    api "org.jetbrains.kotlin:kotlin-test:$kotlin_version"
```
Comment (erikhuizinga) [inline uniflow-test/build.gradle:8]: I don't think that this needs to be a transitive dependency.

In the future you could take a look at https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin to see what dependencies might need being added, removed or changed.

**Review by erikhuizinga — APPROVED** [review]
Functionally approved ✅
Feel free to (not) address my 2 comments.

File: gradle/versions.gradle (hunk 6 above)
Comment (arnaudgiuliani) [reply gradle/versions.gradle:7]: yes somehow :)

File: uniflow-test/build.gradle (hunk 7 above)
Comment (arnaudgiuliani) [reply uniflow-test/build.gradle:8]: as we use in android-test packages, it's interesting to make it transitive for us ... not sure it changes anything in metadata

**Pull request merged**